Add Support for Indexers and Ranges

RexJaeschke opened the pull request:
Before you look at the detailed edits, it likely will be useful to read the following:

- 14.7 Properties|14.7.1 General, which defines ***countable*** as it pertains to a type.
- 14.(x) Indexable sequences, which defines ***indexable sequence***.

RexJaeschke committed:
Add range operator

RexJaeschke committed:
add support for indexers and ranges

RexJaeschke committed:
add indexers and ranges

RexJaeschke committed:
add indexers and ranges

RexJaeschke committed:
add indexers and ranges

RexJaeschke committed:
fix formatting

RexJaeschke committed:
fix formatting

RexJaeschke committed:
fix formatting

RexJaeschke committed:
fix formatting

RexJaeschke committed:
fix formatting

RexJaeschke committed:
fix formatting

RexJaeschke committed:
fix formatting

KalleOlaviNiemitalo commented on standard/classes.md:
```diff
@@ -3049,6 +3049,8 @@ Even though the syntax for accessing a property is the same as that for a field,
 
 When a property declaration includes an `extern` modifier, the property is said to be an ***external property***. Because an external property declaration provides no actual implementation, each of its *accessor_declarations* consists of a semicolon.
 
+A type is ***countable*** if it has a property named `Length` or `Count` with an accessible `get` accessor ([§14.7.3]( classes.md#1473-accessors)) and a return type of `int`.
```
Instance property only, or also static?

RexJaeschke replied:
Instance only.

KalleOlaviNiemitalo commented on standard/classes.md:
```diff
@@ -5087,3 +5089,136 @@ When the body of the async function terminates, the return task is moved out of
+- The type has an accessible instance indexer taking an argument of type `int` as its only argument, or as its first argument with the remaining arguments being optional.
+- The type does not have an accessible instance indexer taking a `System.Index` as its only argument, or as its first argument with the remaining arguments being optional.
+
+The provided instance indexer shall have the same get and set members with matching accessibility as the `int` indexer.
```
The provided instance indexer shall not be virtual.

KalleOlaviNiemitalo replied:
Moreover…

```csharp
class C1 {
    public int this[int index] => 0;
    public int Count => 42;
    
    // provided by the implementation:
    // public int this[System.Index index] => this[index.GetOffset(this.Count)];
}

class C2 : C1 {
    public new int this[System.Index index] => 0;
}

class D {
    static void Main()
    {
        C1 c = new C2();
        _ = c[^0];
    }
}
```

This code should cause a warning

> warning CS0109: The member 'C2.this[Index]' does not hide an accessible member. The new keyword is not required.

even though C2.this[Index] hides the C1.this[Index] indexer that is provided by the implementation.

RexJaeschke replied:
I now require the generated indexer to be non-virtual.

KalleOlaviNiemitalo commented on standard/classes.md:
```diff
@@ -5087,3 +5089,136 @@ When the body of the async function terminates, the return task is moved out of
+
+The provided instance indexer shall have the same get and set members with matching accessibility as the `int` indexer.
+
+The provided instance indexer shall take the given `System.Index` and use that to call the instance indexer taking an `int`. If both the `Length` and `Count` properties exist and are accessible, `Length` is used.
```
Should normative text specify the GetOffset call in more detail?

RexJaeschke replied:
As TG3 (CLI) is inactive, we (TG2) are adding new types to Annex C. However, we only show the signatures, without any semantics. That said, `GetOffset` is only used once normatively, and I extended that use, as shown in bold below:

> For an index expression having type `System.Index`, the Index is transformed into the corresponding `int` index using `System.Index.GetOffset`**, which calculates the offset from the start of the collection using the specified collection length**."

KalleOlaviNiemitalo commented on standard/classes.md:
```diff
@@ -5087,3 +5089,136 @@ When the body of the async function terminates, the return task is moved out of
+  > *Note*: As specified in [§11.7.10.2](expressions.md#117102-array-access), for array access, the method `System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray` is used instead of `Slice`. *end note*
+- The type does not have an accessible instance indexer taking a `System.Range` as its only argument, or as its first argument with the remaining arguments being optional.
+
+The provided instance indexer shall have the same accessibility and return type, including `ref` if present, as `Slice`.
```
Again not virtual.

RexJaeschke replied:
I now require the generated indexer to be non-virtual.

KalleOlaviNiemitalo commented on standard/classes.md:
```diff
@@ -5087,3 +5089,136 @@ When the body of the async function terminates, the return task is moved out of
+An implementation shall behave as if it provides an instance indexer member with a single parameter of type `System.Range` for any type that meets the following criteria:
+
+- The type is countable [§14.7.1](classes.md#1471-general).
+- The type has an accessible instance method named `Slice` taking two arguments of type `int` as the only arguments. For type `string`, the method `Substring` is used instead of `Slice`.
```
OK, does not search for a Slice extension method.

RexJaeschke commented:
[I raised the questions below on https://github.com/dotnet/docs/issues/30512, and got replies, which I need to factor in to this Draft PR.] 

I'm transforming this proposal into a Draft PR for the C# Standard, and have several questions:

**Q1. Optional Members in Index and Range**: The proposal suggests that some members in these types are optional, such as Range's StartAt, EndAt, and All. Which members are required for these types? Why not require all those provided by MS's implementation?

Specifically, the MS proposal states:

> The `..` syntax for `System.Range` will require the `System.Range` type, as well as one or more of the following members:
>
> ```csharp
> namespace System
> {
>     public readonly struct Range
>     {
>         public Range(System.Index start, System.Index end);
>         public static Range StartAt(System.Index start);
>         public static Range EndAt(System.Index end);
>         public static Range All { get; }
>     }
> }
> ```
> 
> The `..` syntax allows for either, both, or none of its arguments to be absent. Regardless of the number of arguments, the `Range` constructor is always sufficient for using the `Range` syntax. However, if any of the other members are present and one or more of the `..` arguments are missing, the appropriate member may be substituted.

MS's feedback on my Q was:

> The only one required is the constructor: we will fall back to it if any of the others are missing. As to why it was specified this way, I'm not certain, and unfortunately Andy (who implemented it) just went on leave for a few months.

As a result, I'm omitting any mention of the optional members `StartAt`, `EndAt`, and `All`, and defining things in terms of the constructor.


**Q2. Range indexer setter**: What if anything should we say about the implicit and explicit setter for a Range indexer? Certainly, one can define a setter for a user-defined type; however, it is not obvious as to what such a setter would do, especially since it must be used on the left-hand side of assignment taking a right-hand side of the same type as the index returns. In the case of type BitArray that would mean something like ba1[range1] = ba2, or perhaps ba1[range1] = ba2[range2]. As far as I can determine, the operations one might like to implement using such a setter are probably best implemented via a named method. In any event, for a compiler-generated Range indexer, attempting to use its setter results in the error message “CS0131 The left-hand side of an assignment must be a variable, property or indexer,” which suggests the generated indexer has no setter. If that is the case, we should say that, perhaps by stressing that the result of a Range indexer is not a variable, so as such, it can't be used on the lhs of an assignment.

MS's feedback on my Q was:

> I think we should say nothing about it, or at most say that no setter is synthesized if one does not exist. I think the rest of the rules should just fall out from this, as the error implies.

I chose to say nothing about the setter, which means that attempting to use such a setter results in unspecified behavior.

RexJaeschke commented:
When adapting the MS proposal, I invented the term ***indexable sequence***. The rationale for that name choice follows.

Various MS-hosted on-line pages use the term *sequence*. This word is already used quite a bit in the C# spec, in both a general sense as well as being defined in the context of query expressions. §11.17 Query expressions|§11.17.1 General states:

> A query expression begins with a `from` clause and ends with either a `select` or `group` clause. The initial `from` clause may be followed by zero or more `from`, `let`, `where`, `join` or `orderby` clauses. Each `from` clause is a generator introducing a ***range variable*** that ranges over the elements of a ***sequence***. Each `let` clause introduces a range variable representing a value computed by means of previous range variables. ….

That definition is *not* applicable to indexes and ranges! The MS-provided proposal uses *collection*; however, that implies enumerable support, which is *not* required by indexes and ranges. (BTW, although it is used a lot in the C# spec, the term *collection* is *not* defined!) As such, rather than overload an existing term or invent a completely different one, I came up with *indexable sequence*.

RexJaeschke commented:
The [MS proposal](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md), section “Implicit Range support,” provided a very detailed discussion of how to transform a pair of Indexes into a call to `Slice` depending on the form of the range used. I did *not* retain this in the final proposal, as I saw no point in doing so. Given a start and end index, it is a simple matter to compute the length **in all cases regardless of range format!**, as I show in my range indexer implementation in “Explicit range support.”

RexJaeschke committed:
fix md formatting

RexJaeschke committed:
fix md formatting

RexJaeschke committed:
fix md formatting

RexJaeschke committed:
fix md formatting

RexJaeschke committed:
fix md formatting

RexJaeschke committed:
Add annotation to examples

RexJaeschke committed:
Create BitArrayPartial1.cs

RexJaeschke committed:
Create BitArrayPartial2.cs

RexJaeschke committed:
Create BitArrayPartial3.cs

RexJaeschke committed:
added annotation to examples

RexJaeschke committed:
fix md formatting

BillWagner commented:
rebased on the latest draft-v8 branch on 09/26/2023

BillWagner committed:
update links

BillWagner committed:
more section references

BillWagner committed:
more section numbers

RexJaeschke committed:
Small bits of cleanup

RexJaeschke committed:
Make some small improvements

RexJaeschke committed:
Remove non-required members

RexJaeschke committed:
Merge branch 'draft-v8' into indexers-and-ranges

RexJaeschke committed:
fix type name ordering

RexJaeschke committed:
describe GetOffset

RexJaeschke committed:
some tweaks

- added non-virtual requirement to 2 x imp-generated indexers.
- described `GetOffset` semantics.
- removed erroneous text w.r.t optional trailing params on `this[int]`.

RexJaeschke committed:
Merge branch 'draft-v8' into indexers-and-ranges

RexJaeschke commented:
@KalleOlaviNiemitalo I just revisited this PR and resolved all of Kalle's comments. While doing that, I found a requirement I'd added that was incorrect, so I removed that.

jnm2 commented on standard/classes.md:
```diff
@@ -3244,6 +3244,8 @@ In a *ref_property_body* an expression body consisting of `=>` followed by `ref`
 
 When a property declaration includes an `extern` modifier, the property is said to be an ***external property***. Because an external property declaration provides no actual implementation, each of the *accessor_body*s in its *accessor_declarations* shall be a semicolon.
 
+A type is ***countable*** if it has an instance property named `Length` or `Count` with an accessible `get` accessor ([§15.7.3]( classes.md#1573-accessors)) and a return type of `int`.
```
Question: is this a place where the spec needs to be specific about `int` versus other ways to reference the type (e.g. `global::System.Int32` as well as aliases)?

Also, nit:
```suggestion
A type is ***countable*** if it has an instance property named `Length` or `Count` with an accessible `get` accessor ([§15.7.3](classes.md#1573-accessors)) and a return type of `int`.
```

BillWagner replied:
> Question: is this a place where the spec needs to be specific about `int` versus other ways to reference the type (e.g. `global::System.Int32` as well as aliases)?

That should be covered under 8.3.5 (Simple types)

jskeet replied:
Resolving as covered.

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+An index is represented by a read-only variable of the value type `System.Index`. A range is represented by a read-only variable of value type `System.Range`, which contains a start and end index. A slice of an array is represented by a (possibly empty) array. The representation of a slice of a user-defined type is determined by the implementer of that type.
+
+For an indexable sequence of length *N*, elements can be accessed using indexes 0 through *N-1*, which are relative to the start. Elements can also be accessed relative to the end via the `^`  index-from-end operator (§index-from-end-operator). `^0` denotes the (non-existent) element just beyond the end.
```
We could specify here the range of valid values that may be provided with the `^` operator, similar to the "0 through N-1" text earlier on the line. In this case it would be "1 through N."

Also, nit:
```suggestion
For an indexable sequence of length *N*, elements can be accessed using indexes 0 through *N-1*, which are relative to the start. Elements can also be accessed relative to the end via the `^` index-from-end operator (§index-from-end-operator). `^0` denotes the (non-existent) element just beyond the end.
```

Nigel-Ecma replied:
@jnm2 – 0 through N for `^` to allow for `^0`? Otherwise 👍

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+For an indexable sequence of length *N*, elements can be accessed using indexes 0 through *N-1*, which are relative to the start. Elements can also be accessed relative to the end via the `^`  index-from-end operator (§index-from-end-operator). `^0` denotes the (non-existent) element just beyond the end.
+
+A slice can be obtained using the `..` range operator (§range-operator). A range of the form `s..e` starts at element `s` and ends with the element immediately prior to element `e`.
```
We could mention here that either or both ends may be omitted, each referring then to its appropriate extremity within the available range.

BillWagner replied:
I think it's fine as is. The other forms of `..` are in the linked section.

Nigel-Ecma replied:
No, a slice might be obtained using a range – or it might fail, a range can be constructed using the range operator…

TBD: Wordsmith a suggestion

Nigel-Ecma replied:
The TBD from the preceding comment:

```suggestion
A `Range` value can be obtained using the `..` range operator (§range-operator). An expression of the form `s..e` produces a `Range` value which starts at index`s` and ends with the index immediately prior to index `e`.

A slice of an indexable collection can be obtained by indexing the collection using a `Range` value.
```

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+A slice can be obtained using the `..` range operator (§range-operator). A range of the form `s..e` starts at element `s` and ends with the element immediately prior to element `e`.
+
+All single-dimensional and jagged arrays ([§17.1](arrays.md#171-general)) are indexable sequences; multi-dimensional arrays are not! The use of indexes and ranges with arrays is described in [§12.8.11.2](expressions.md#128112-array-access).
```
Is it typical in the spec to mention jagged arrays as though they were separate from single-dimensional arrays? I had thought of jagged arrays as a special case of single-dimensional arrays.

BillWagner replied:
I might consider it a note:

```suggestion
All single-dimensional and jagged arrays ([§17.1](arrays.md#171-general)) are indexable sequences; multi-dimensional arrays are not! The use of indexes and ranges with arrays is described in [§12.8.11.2](expressions.md#128112-array-access).

> *Note*: Jagged arrays ([§17.1](arrays.md#171-general)) are arrays of single-dimensional arrays, and are therefore indexable. *end note*.
```

Nigel-Ecma replied:
I agree with @jnm2 here – C# doesn’t have jagged arrays *per se*, it supports single-dimensional arrays where the element type can be an array which:

> can be used to represent “jagged arrays”. [§17.1]

I would just drop the reference (and the `!`), x-ref fixed as per @BillWagner comment below:

```suggestion
All single-dimensional are indexable sequences; multi-dimensional arrays are not. The use of indexes and ranges with arrays is described in [§12.8.12.2](expressions.md#128122-array-access).
```

jskeet replied:
"All single-dimensional are indexable sequences" - this needs to be "All single-dimensional arrays are indexable sequences"

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+An object of type `string` is an indexable sequence.
+
+A user-defined type can provide explicit support for indexer access ([§12.8.11.3](expressions.md#128113-indexer-access)) using `System.Index` and `System.Range`. (See §indexable-sequence-expl-support-for-index and §indexable-sequence-expl-support-for-range.) If various criteria are met, an existing user-defined type that does *not* have such explicit support, shall have provided for it by the imp ...
```
One sentence felt hard for me to follow. Does this version achieve the same goals you're looking to achieve?
```suggestion
A user-defined type can provide explicit support for indexer access ([§12.8.11.3](expressions.md#128113-indexer-access)) using `System.Index` and `System.Range`. (See §indexable-sequence-expl-support-for-index and §indexable-sequence-expl-support-for-range.) If various criteria are met, the implementation shall provide the same support for such indexer and range access on an existing user-defined type that does *not* have such explicit support. (See §indexable-sequence-impl-support-for-index and §indexable-sequence-impl-support-for-range.) In both cases, the type is recognized as being an indexable sequence type.
```

BillWagner replied:
I like Joseph's wording here.

Nigel-Ecma replied:
For convenience updating @jnm2’s suggestion with @BillWagner x-ref update:

```suggestion
A user-defined type can provide explicit support for indexer access ([§12.8.1123](expressions.md#128123-indexer-access)) using `System.Index` and `System.Range`. (See §indexable-sequence-expl-support-for-index and §indexable-sequence-expl-support-for-range.) If various criteria are met, the implementation shall provide the same support for such indexer and range access on an existing user-defined type that does *not* have such explicit support. (See §indexable-sequence-impl-support-for-index and §indexable-sequence-impl-support-for-range.) In both cases, the type is recognized as being an indexable sequence type.
```

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+#### §indexable-sequence-expl-support-for-index Explicit Index support
+
+A type having an instance indexer taking a single argument of type `System.Index`, or a first argument of that type followed by optional arguments, may be indexed as described by [§12.8.11.3](expressions.md#128113-indexer-access).
```
I may be off on a tangent here, and I'm happy to get a better understanding of what the aims are in this sentence, but here goes 😁

If optional arguments are mentioned, is `params` implicitly considered an optional argument, and if not, should it be mentioned alongside optional arguments here?

Also, the Roslyn compiler is happy with this code, which does not have an argument of type `System.Index`:

```cs
using System;

var expr = new C();

_ = expr[^2];
_ = expr is [];

class C
{
    public object this[params Index[] x] => throw null;

    public int Count => throw null;
}
```

In addition, due to implicit conversions, the type need not be `System.Index` so long as overload resolution succeeds:

```cs
using System;

var expr = new C();

_ = expr[^2];
_ = expr is [];

class C
{
    public object this[params CustomIndex[] x] => throw null;

    public int Count => throw null;
}

class CustomIndex
{
    public static implicit operator CustomIndex(Index i) => throw null;
}
```

The compiler is also happy with the indexer having a `dynamic` parameter, etc.

We also didn't mention refness of the parameter (`in` is allowed, but `ref` and `out` are not), etc.

I've observed the compiler team defining this general concept in terms of being able to bind arguments to an applicable overload. If we defined things the same way in the spec somewhere, we could refer to that definition. Thereby maybe we could sidestep all the gotchas around repeating ourselves exhaustively, as to what allows binding to succeed for these features.

There are other instances below to which this would also apply.

BillWagner replied:
Could the change be this simple:

```suggestion
A type having an instance indexer taking a single argument with an implicit conversion to `System.Index`, or a first argument with an implicit conversion to `System.Index` followed by optional arguments, may be indexed as described by [§12.8.11.3](expressions.md#128113-indexer-access).
```

Nigel-Ecma replied:
> I may be off on a tangent here, and I'm happy to get a better understanding of what the aims are in this sentence, but here goes 😁

Tangents are welcome! A core part of the task is looking into all the nooks’n’crannies that might not be covered in the source material.

> If optional arguments are mentioned, is `params` implicitly considered an optional argument, and if not, should it be mentioned alongside optional arguments here?
> 
> Also, the Roslyn compiler is happy with this code, which does not have an argument of type `System.Index`:

How an indexer access is resolved to a particular indexer is already covered in §12.8.11.3 *Indexer access* which in turn relies on §12.6.4 *Overload resolution* – between them these cover such things as optional parameters, `params` and implicit conversions. So this para is, incompletely as you have identified, echoing part of the text it is actually referencing.

So rather specifying indexers that have an argument of type `Index`, we need to specify in terms of indexers that *accept*/*can be applied to* a single argument of type `Index` and then refer to §12.8.11.3 which covers all the cases.

Suggested rewording:

```suggestion
A type having an instance indexer which can be applied to a single argument of type `System.Index` may be indexed as described by [§12.8.11.3](expressions.md#128113-indexer-access).
```

Once the para is reduced to this it is clear it is stating the blindly obvious – before `System.Index` you could write an indexer taking any type of argument you wish, after you still can. Do we really need this at all?

Following below there may be further suggestions changing “take” to ”accept”/”apply” – this and those need to be accepted/rejected as a whole.

jskeet replied:
Aside from anything else: I think these should refer to parameters, not arguments. There's no such thing as an optional argument, only an optional parameter.

Nigel-Ecma replied:
@jskeet – I think the use of “argument” is correct in my suggestion?

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+- The type has an accessible instance indexer taking an argument of type `int` as its only argument.
+- The type does not have an accessible instance indexer taking a `System.Index` as its only argument, or as its first argument with the remaining arguments being optional.
+
+The provided instance indexer shall have the same get and set members with matching accessibility as the `int` indexer.
```
Does the spec generally need to talk about `init` accessors when `set` accessors are mentioned?

BillWagner replied:
`init` was added in C# 9.0 

We can defer that discussion!

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+#### §indexable-sequence-impl-support-for-index Implicit Index support
+
+An implementation shall behave as if it provides a non-virtual instance indexer member with a single parameter of type `System.Index` for any type that meets the following criteria:
```
This could be clarified. The implementation only behaves this way in regard to the call site which calls this synthesized indexer. It does not, for instance, behave as if it provides this indexer within the class definition itself when the class is being compiled:

```cs
Console.WriteLine(typeof(C).GetProperties().Length); // 2

class C
{
    public int this[int index] { get => throw null; set => throw null; }
    public int Count => throw null;
}
```

Same again below.

Nigel-Ecma replied:
@jnm2 – Hmm, I read “shall behave *as if* it provides” (emphasis added) as clearly stating what you’d like clarified – but folk do read things differently. Can you suggest some wording?

BTW I would expect that an implementation *could* choose to add an actual indexer to the type should it wish and be compliant, so if folk agree and don’t think this wording allows that then some wordsmithing will be required.

jskeet replied:
I'd personally *not* expect it to add an actual indexer to the type, as that could then be detected as a breaking change later if someone adds an *actual* indexer with a different parameter name.

But I do think it could potentially be clarified in terms of what's doing the "providing", e.g.

> For any type `T` that meets the following criteria, an implementation shall behave as if `T` provides a non-virtual instance indexer member with a single parameter of type `System.Index`

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+A type having an instance indexer taking a single argument of type `System.Range`, or a first argument of that type followed by optional arguments, may be indexed as described by [§12.8.11.3](expressions.md#128113-indexer-access).
+
+> *Example*: In [§15.9](classes.md#159-indexers), there is an example defining type `BitArray`, which stores bits in an array of `int`. Adding a `Range` indexer that returns a `BitArray` representing the bit slice designated by the Range, is simple:
```
"Simple" seems to be a subjective term. What would the goal be to communicate here?

Nigel-Ecma replied:
@jnm2 – maybe “straightforward” which doesn’t carry the same risk of offense as “simple”, or may be:

```suggestion
> *Example*: In [§15.9](classes.md#159-indexers), there is an example defining type `BitArray`, which stores bits in an array of `int`. Adding a `Range` indexer that returns a `BitArray` representing the bit slice designated by the `Range` may be achieved as follows:
```

Wordsmith away folks!

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+An implementation shall behave as if it provides a non-virtual instance indexer member with a single parameter of type `System.Index` for any type that meets the following criteria:
+
+- The type is countable ([§15.7.1](classes.md#1571-general)).
+- The type has an accessible instance indexer taking an argument of type `int` as its only argument.
```
In contrast to another comment thread, this is an example where "able to bind `(int)`" is _not_ the only requirement, since subsequent optional parameters _do_ throw it off. :+1:

Is parameter refness important to mention? `in int` is not supported, for instance.

Nigel-Ecma replied:
[Replaced] (My fears were justified)

For the refness question, maybe:

> The type has an accessible instance indexer taking a single parameter of type `int` without any associated *parameter_mode_modifier*.

jskeet replied:
Again, I think we should talk about the declared parameters, rather than "taking an argument".

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+An implementation shall behave as if it provides a non-virtual instance indexer member with a single parameter of type `System.Range` for any type that meets the following criteria:
+
+- The type is countable ([§15.7.1](classes.md#1571-general)).
+- The type has an accessible instance method named `Slice` taking two arguments of type `int` as the only arguments. For type `string`, the method `Substring` is used instead of `Slice`.
+  > *Note*: As specified in [§12.8.11.2](expressions.md#128112-array-access), for array access, the method `System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray` is used instead of `Slice`. *end note*
```
This is listed as a note, and arrays are left out of the list on the previous line. It seems like arrays should be given equal special-case status as strings. These are the two types that are the exception to needing a method named Slice.

If the Roslyn compiler can't find the method RuntimeHelpers.GetSubArray, it still behaves as if it provides the range indexer, and then fails at a later stage with "CS0656 Missing compiler required member 'System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray'." So the presence of this method is not a determining factor for whether the Roslyn implementation behaves as if it provides the Range indexer.

I didn't check what happens if `System.String` doesn't have a `Substring` method because that would take longer to test, but perhaps the logic is this:

```md
- The type is `string`, an array type, or has an accessible instance method named `Slice` [...].
```

Then, mentions of the Substring and GetSubArray methods could be notes informing readers what will happen in those two special cases.

Nigel-Ecma replied:
If a different method is being used for arrays then it needs to be specified in normative text, not as an informative note.

jskeet replied:
And we need to think about whether `GetSubArray` is *required*, or whether it would be okay for an implementation to actually provide a method called `Slice` instead...

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+  > *Note*: As specified in [§12.8.11.2](expressions.md#128112-array-access), for array access, the method `System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray` is used instead of `Slice`. *end note*
+- The type does not have an accessible instance indexer taking a `System.Range` as its only argument, or as its first argument with the remaining arguments being optional.
+
+The provided instance indexer shall have the same accessibility and return type, including `ref` if present, as `Slice`.
```
Should `ref readonly` be mentioned too?

Nigel-Ecma replied:
Use *ref_kind*, I suspect it just predates this?

```suggestion
The provided instance indexer shall have the same accessibility and return type, including any *ref_kind* if present, as `Slice`.
```

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+- The type has an accessible instance indexer taking an argument of type `int` as its only argument.
+- The type does not have an accessible instance indexer taking a `System.Index` as its only argument, or as its first argument with the remaining arguments being optional.
+
+The provided instance indexer shall have the same get and set members with matching accessibility as the `int` indexer.
```
The 'range' version of this called out that the return type was the same, including 'ref'. That is true here too.

jnm2 commented on standard/classes.md:
```diff
@@ -5481,3 +5483,129 @@ When the body of the async function terminates, the return task is moved out of
+
+The provided instance indexer shall have the same accessibility and return type, including `ref` if present, as `Slice`.
+
+When the type is indexed with a `System.Range`, the provided instance indexer shall take the given range and pass its start index and length as `int`s to `Slice` (or in the case of `string`, to method `Substring`).
```
This sentence doesn't hold for arrays.

RexJaeschke committed:
Merge branch 'draft-v8' into indexers-and-ranges

RexJaeschke committed:
Add new types to C.5

BillWagner commented on standard/standard-library.md:
```diff
@@ -618,7 +644,12 @@ namespace System.Runtime.CompilerServices
     }
 
-    public readonly struct TaskAwaiter : ICriticalNotifyCompletion,
+    public static class RuntimeHelpers
+    {
+        public static T[] GetSubArray<T>(T[] array, System.Range range);
+    }
```
I'll check the normative language as well, but this type seems to be an implementation specific choice. Are we going to standardize that this is the method to retrieve a range in an array?

jskeet replied:
Call on 2025-02-19: let's not specify this, but instead, where it's currently referenced, we should specify the result. We don't want this to be "implementation-defined" (as we don't want implementations to give their methods). We will want to make it clear that it *is* a shallow copy, and permit but not require that the same reference to an empty array can be returned by multiple calls.

BillWagner commented on standard/standard-library.md:
```diff
@@ -379,18 +379,44 @@ The following types, including the members listed, shall be defined in a conform
 
 A conforming implementation may provide `Task.GetAwaiter()` and `Task<TResult>.GetAwaiter()` as extension methods.
 
+> Note to TG2 reviewers: Required vs. Optional library type members: We need to indicate which members of `Index` and `Range` are required and which are optional.
```
My first thought is to keep it simple. I'd just leave them all. I don't have a super strong opinion, but that's my first thought.

jskeet replied:
Just for comparison, here's what we say about `String`:

```csharp
    public sealed class String : IEnumerable<Char>, IEnumerable
    {
        public int Length { get; }
        public char this [int index] { get; }
        public static string Format(string format, params object?[] args);
    }
```

We don't define `IEquatable<T>` anywhere, so arguably we shouldn't use it (although that's already present for `ValueTask`)

jskeet replied:
C.1 states: "The standard library is intended to be the minimum set of types and members required by a conforming C# implementation. As such, it contains only those members that are explicitly required by the C# language specification."

jskeet replied:
But if everyone else is happy being maximal, let's do that.

BillWagner commented on tools/example-templates/additional-files/BitArrayPartial1.cs:
```diff
@@ -1,66 +1,66 @@
+        ba1[0] = true;
+        ba1[^98] = true;
+        ba1[99] = true;
+        Console.WriteLine("ba1[0] = {0}", ba1[0]);                    // True
+        Console.WriteLine("ba1[98] = {0}", ba1[98]);                  // False
+        Console.WriteLine("ba1[Index 0] = {0}", ba1[new Index(0)]);   // True
+        Console.WriteLine("ba1[^1] = {0}", ba1[^1]);                  // True
```
Since we have them, I'd use interpolation expressions:

```suggestion
        Console.WriteLine($"ba1[0] = {ba1[0]}");                    // True
        Console.WriteLine($"ba1[98] = {ba1[98]}");                  // False
        Console.WriteLine($"ba1[Index 0] = {ba1[new Index(0)]}");   // True
        Console.WriteLine($"ba1[^1] = {ba1[^1]}");                  // True
```

BillWagner commented on tools/example-templates/additional-files/BitArrayPartial2.cs:
```diff
@@ -1,115 +1,115 @@
+        ba[0] = true;
+        ba[3] = true;
+        ba[^1] = true;
+        Console.WriteLine("ba = >{0}<", ba);
```
Same comment:

```suggestion
        Console.WriteLine($"ba = >{ba}<");
```

BillWagner commented on tools/example-templates/additional-files/BitArrayPartial3.cs:
```diff
@@ -1,116 +1,116 @@
+        ba[0] = true;
+        ba[3] = true;
+        ba[^1] = true;
+        Console.WriteLine("ba = >{0}<", ba);
```
again:

```suggestion
        Console.WriteLine($"ba = >{ba}<");
```

BillWagner commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,99 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
-- The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
-- The value of each expression in the *argument_list* is checked against the actual bounds of each dimension of the array instance referenced by `P`. If one or more values are out of range, a `System.IndexOutOfRangeException` is thrown and no further steps are executed.
-- The location of the array element given by the index expression(s) is computed, and this location becomes the result of the array access.
+- The index expressions of the *argument_list* are evaluated in order, from left to right. Following evaluation of each index expression, for expressions not having type `System.Index` or `System.Range`, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to one of the following types is performed: `int`, `uint`, `long`, `ulong`. The first type in this list for which an impl ...
```
I *think* Index should be code fenced, if it's capitalized.  (Alternative: don't capitalize it):

```suggestion
- The index expressions of the *argument_list* are evaluated in order, from left to right. Following evaluation of each index expression, for expressions not having type `System.Index` or `System.Range`, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to one of the following types is performed: `int`, `uint`, `long`, `ulong`. The first type in this list for which an implicit conversion exists is chosen. For instance, if the index expression is of type `short` then an implicit conversion to `int` is performed, since implicit conversions from `short` to `int` and from `short` to `long` are possible. For an index expression having type `System.Index`, the `Index` is transformed into the corresponding `int` index using `System.Index.GetOffset`, which calculates the offset from the start of the collection using the specified collection length.
```

BillWagner commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,99 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+- The index expressions of the *argument_list* are evaluated in order, from left to right. Following evaluation of each index expression, for expressions not having type `System.Index` or `System.Range`, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to one of the following types is performed: `int`, `uint`, `long`, `ulong`. The first type in this list for which an impl ...
+
+- For an index expression not having type `System.Range`:
+  - If evaluation of an index expression, the subsequent implicit conversion, or Index transformation causes an exception, then no further index expressions are evaluated, and no further steps are executed.
```
Why is "Index" capitalized here?

Nigel-Ecma replied:
Maybe both Index here and Range below are meant to be the types `Index` & `Range`?

BillWagner commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,99 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+  - The location of the array element given by the index expression(s) is computed, and this location becomes the result of the array access.
+- Else, for an index expression having type `System.Range`:
+  - The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
+  - The Range is transformed into the corresponding slice using `System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray<T>`. If this transformation causes an exception, no further steps are executed.
```
See comment in standard-library.md.  Do we want to make it the standard that `GetSubArray` is how this happens?  One alternative:

```suggestion
  - The Range is transformed into the corresponding slice of the same *array_type* as `P`. If this transformation causes an exception, no further steps are executed.
```

BillWagner commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,99 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+> int[,] values2D = {{10, 5, 7, 1}, {34, 13, 6, 2}};
+> Index idx3 = 1; 
+> Index idx4 = ^1;
+> Console.WriteLine(values2D[idx3, idx4]);      // won't compile!
```
Should we use more standard like language here?

```suggestion
> Console.WriteLine(values2D[idx3, idx4]);      // compiler error.
```

BillWagner commented on standard/expressions.md:
````diff
@@ -2227,17 +2228,99 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+> string[] names = seasons[0..2]; // slice containing "Summer" and "Autumn"
+> ```
+>
+> `seasons[0..2]` is transformed by the implementation to
+>
+> <!-- NotAn$Example: {} -->
+> ```csharp
+>  System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray<string>(seasons, 0..2)
+> ```
+>
+> which returns a `string[]` slice.
````
If we decide that `GetSubArray` isn't required by the standard, this could become:

```suggestion
> `seasons[0..2]` is transformed by the implementation return a `string[]` slice that contains 2 elements.
```

BillWagner commented on standard/expressions.md:
````diff
@@ -2227,17 +2228,99 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+> Console.WriteLine(values[1..3][^1][..2][^1]);      // 42
+> ```
+>
+> all the Range and Index expressions in `values[1..3][^1][..2][^1]` are transformed by the implementation, resulting in `values[2][1]`, which is 42. *end example*
````
Because this is an example, let's consider writing out each transformation:  `values[1..3]` evaluates to .... Then `result`[^1]` evaluates to `result2` which is .... Then, `result2[..2]` evaluates to `result3` which is .... Finally, `result3[^1]` is 42.

BillWagner commented on standard/expressions.md:
```diff
@@ -2256,6 +2339,8 @@ The binding-time processing of an indexer access of the form `P[A]`, where `P` i
 
 Depending on the context in which it is used, an indexer access causes invocation of either the get accessor or the set accessor of the indexer. If the indexer access is the target of an assignment, the set accessor is invoked to assign a new value ([§12.21.2](expressions.md#12212-simple-assignment)). In all other cases, the get accessor is invoked to obtain the current value ([§12.2.2](expressio ...
 
+> *Note*: An implementation is required to provide an instance indexer member with a single parameter of type `Index` for any type that meets the criteria specified in §indexable-sequence-impl-support-for-index. An implementation is required to provide an instance indexer member with a single parameter of type `Range` for any type that meets the criteria specified in §indexable-sequence-impl-supp ...
```
Because this is *required*, should it be normative? Should we state that the compiler must translate those expressions into the equivalent expressions using the integral types?

jskeet replied:
I think this should probably be rewritten anyway, in terms of "behaving as if an instance indexer member is provided".
(Again, this is up for discussion - basically, do we expect that to be in the emitted IL, like a default instance constructor is? I suspect not.)

Nigel-Ecma replied:
This *Note* contradicts the clauses it references. With @jskeet as to the need to rewrite (normatively) here. Whether it describes “as if an indexer” exists (as in the referenced clauses), an argument conversion (as if) some call to `GetOffset()` is called (as in the changes currently suggested to array indexing), or as some kind of “target type conversion” (as [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#adding-index-and-range-support-to-existing-library-types)) is up for discussion. Hopefully we can describe array and non-array indexing behaviour very similarly.

BillWagner commented on standard/expressions.md:
```diff
@@ -3557,6 +3643,47 @@ The result of evaluating `~x`, where `X` is an expression of an enumeration ty
 
 Lifted ([§12.4.8](expressions.md#1248-lifted-operators)) forms of the unlifted predefined bitwise complement operators defined above are also predefined.
 
+### §index-from-end-operator Index-from-end operator
```
Somewhere in this section (perhaps in a note), we should state that `^0` represents the end of the sequence, not the last element. `^1` represents the last element. (Yes, I know it's in a comment in the example, but still...)

Nigel-Ecma replied:
I think that should be explained where `System.Index` is introduced or here, with x-refs between them.

BillWagner commented on standard/expressions.md:
````diff
@@ -3688,6 +3815,55 @@ At run-time, the expression `await t` is evaluated as follows:
+This operator provides a succinct syntax for specifying a (possibly empty) element range suitable for use in denoting a slice of an indexable sequence (§indexable-sequence).
+
+```ANTLR
+range_expression
+    : unary_expression
+    | range_expression? '..' range_expression?
+    ;
````
This is triggering the left recursive issue in the grammar. I defer to @Nigel-Ecma for a suggestion on an alternative expression.

BillWagner replied:
Also, I don't think this grammar is correct. Would it permit something like `P[0..5..8..^2]`?

Nigel-Ecma replied:
Agree, the grammar is wrong, see later comment.

BillWagner commented on standard/expressions.md:
```diff
@@ -3688,6 +3815,55 @@ At run-time, the expression `await t` is evaluated as follows:
+
+The left and right operands denote, respectively, a start and end Index. For this operator, an object of (the immutable struct) type `System.Range` is returned that contains those Indexes. If the left operand is omitted, an Index of `0` is used. If the right operand is omitted, an Index of `^0` is used. As such,
+
+- `s .. e` is transformed by the implementation to `new System.Range(s, e)`.
+- `s ..` is transformed by the implementation to `new System.Range(s, ^0)`.
+- `.. e` is transformed by the implementation to `new System.Range(0, e)`.
```
style nit. Let's omit the spaces here:

```suggestion
- `s..e` is transformed by the implementation to `new System.Range(s, e)`.
- `s..` is transformed by the implementation to `new System.Range(s, ^0)`.
- `..e` is transformed by the implementation to `new System.Range(0, e)`.
```

BillWagner reviewed (commented):
Reviewed everything but classes.md.

BillWagner commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+An ***indexable sequence*** is an ordered set of zero or more elements having the same type. Any given element can be accessed via an index, and a contiguous subset of elements—referred to as a ***slice***—can be denoted via a range.
+
+An index is represented by a read-only variable of the value type `System.Index`. A range is represented by a read-only variable of value type `System.Range`, which contains a start and end index. A slice of an array is represented by a (possibly empty) array. The representation of a slice of a user-defined type is determined by the implementer of that type.
```
Hmm. Should this be a countable type?

```suggestion
An index is represented by a read-only variable of the value type `System.Index`. A range is represented by a read-only variable of value type `System.Range`, which contains a start and end index. A slice of an array is represented by a (possibly empty) array. The representation of a slice of a countable type that is not an array type is determined by the implementer of that type.
```

jskeet replied:
Does a slice have to be of a countable type? You can write an uncountable type with an indexer accepting a range, and make it fail at execution time if either end of the range is "from the end".

I agree it will usually be of a countable type, just trying to explore...

Nigel-Ecma replied:
@jskeet – In C# terms a slice made from a `Range` has to be of a countable type as `Range` is built upon `Index`, and `Index` requires a length for it to be converted into an integer offset…However a non-countable type could provide the same underlying `Slice` method…

Looking at your query made me think how careful we have to be with terms here.

> An index is represented by a read-only variable of the value type `System.Index`.

No, an indexer takes an *index* as its argument, which does not need to be of type `Index`, or even `int`. So “…index **is** represented by…” is wrong, we’re using “index” to both mean a value of any type used as an index and to mean the (`int`, `bool`) pair embodied by `System.Index`…

I think we might yet need a lot of careful wordsmithing. Drat!

BillWagner commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+A user-defined type can provide explicit support for indexer access ([§12.8.11.3](expressions.md#128113-indexer-access)) using `System.Index` and `System.Range`. (See §indexable-sequence-expl-support-for-index and §indexable-sequence-expl-support-for-range.) If various criteria are met, an existing user-defined type that does *not* have such explicit support, shall have provided for it by the imp ...
+
+### §indexable-sequence-support-for-index Providing support for Index
```
For both this section and the corresponding ranges section, I propose we rework them to introduce a pattern based description, following the format used for `async` return types, LINQ, and so on. 

The concept is that a *countable* sequence has a certain pattern for indexer, Count, Length, etc.

A type can provide support for ranges and index operations by implementing a pattern:  Indexer that takes `System.Index` or `in System.Index` etc. `Slice` method that takes a `System.Range`. Then describe ref variants.

Finally, for arrays and other types the compiler recognizes as *countable*, the compiler can synthesize the lowering of the indexer operations.

BillWagner reviewed (commented):
Added comments on the classes.md spec updates.

BillWagner commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+A slice can be obtained using the `..` range operator (§range-operator). A range of the form `s..e` starts at element `s` and ends with the element immediately prior to element `e`.
+
+All single-dimensional and jagged arrays ([§17.1](arrays.md#171-general)) are indexable sequences; multi-dimensional arrays are not! The use of indexes and ranges with arrays is described in [§12.8.11.2](expressions.md#128112-array-access).
```
Array access is now 12.8.12.2:

```suggestion
All single-dimensional and jagged arrays ([§17.1](arrays.md#171-general)) are indexable sequences; multi-dimensional arrays are not! The use of indexes and ranges with arrays is described in [§12.8.12.2](expressions.md#128122-array-access).
```

BillWagner commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+An object of type `string` is an indexable sequence.
+
+A user-defined type can provide explicit support for indexer access ([§12.8.11.3](expressions.md#128113-indexer-access)) using `System.Index` and `System.Range`. (See §indexable-sequence-expl-support-for-index and §indexable-sequence-expl-support-for-range.) If various criteria are met, an existing user-defined type that does *not* have such explicit support, shall have provided for it by the imp ...
```
indexer access moved:

```suggestion
A user-defined type can provide explicit support for indexer access ([§12.8.12.3](expressions.md#128123-indexer-access)) using `System.Index` and `System.Range`. (See §indexable-sequence-expl-support-for-index and §indexable-sequence-expl-support-for-range.) If various criteria are met, an existing user-defined type that does *not* have such explicit support, shall have provided for it by the implementation implicit support for such indexer and range access. (See §indexable-sequence-impl-support-for-index and §indexable-sequence-impl-support-for-range.) In both cases, the type is recognized as being an indexable sequence type.
```

BillWagner commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+#### §indexable-sequence-expl-support-for-range Explicit Range support
+
+A type having an instance indexer taking a single argument of type `System.Range`, or a first argument of that type followed by optional arguments, may be indexed as described by [§12.8.11.3](expressions.md#128113-indexer-access).
```
section change:

```suggestion
A type having an instance indexer taking a single argument of type `System.Range`, or a first argument of that type followed by optional arguments, may be indexed as described by [§12.8.12.3](expressions.md#128123-indexer-access).
```

BillWagner commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+- The type is countable ([§15.7.1](classes.md#1571-general)).
+- The type has an accessible instance method named `Slice` taking two arguments of type `int` as the only arguments. For type `string`, the method `Substring` is used instead of `Slice`.
+  > *Note*: As specified in [§12.8.11.2](expressions.md#128112-array-access), for array access, the method `System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray` is used instead of `Slice`. *end note*
```
section change:

```suggestion
  > *Note*: As specified in [§12.8.12.2](expressions.md#128122-array-access), for array access, the method `System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray` is used instead of `Slice`. *end note*
```

RexJaeschke committed:
Merge branch 'dotnet:draft-v8' into indexers-and-ranges

RexJaeschke committed:
Add constraint for Index in an object initializer

RexJaeschke commented:
I just added a new commit that prohibits an argument of type `Index` in the *argument_list* of an *initializer_target*.

I am working on the spec for a V13 feature (for which there appears to be no formal MS spec, but see [here](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13#implicit-index-access)). That spec removes a constraint that was apparently added in V8 as a result of the addition of support for `Index`, but which was *not* previously mentioned in this V8 proposal.

Consider the following:

```csharp
    public class Program
    {
        public static void Main()
        {
            Index idx = 0;
            var c = new C()
            {
                buffer = { [0] = 8, [1] = 4 }               // V6+ allows
//                buffer = { [(Index)0] = 8, [1] = 4 } // requires V13
//                buffer = { [idx] = 8, [1] = 4 }         // requires V13
//                buffer = { [^1] = 4, [0] = 8 },        // requires V13
            };
        }
    }
}
public class C
{
    public int[] buffer = new int[2];
}
```

V6 added support for the initialization of `buffer` using *initializer_target* notation `[...] = exp`. However, when V8 added support for `System.Index` it ***did not allow the use of a target having that type***, so the 3 commented-out initializations are disallowed. V13 lifts that restriction.

KalleOlaviNiemitalo commented:
> when V8 added support for System.Index it did not allow the use of a target having that type

Does that even disallow initialisation like `new Dictionary<Index, int>() { [^0] = 0 }`, where the Index is not translated?

RexJaeschke commented:
@KalleOlaviNiemitalo Interestingly, your example compiles starting with V8, so ***my recent fix probably needs tweaking***. Let's compare your example with mine:

``` csharp
var d = new Dictionary<Index, int>() { [^0] = 0 };                            // V8+ allows
var c = new C()                      { buffer = { [^1] = 4, [0] = 8 } };      // requires V13
```

In the case of `c`, the `Index` expression is inside a nested *object_initializer*.

Here's the draft-v8 [grammar](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#128173-object-initializers) both are (supposedly) following:

```ANTLR
object_initializer
    : '{' member_initializer_list? '}'
    | '{' member_initializer_list ',' '}'
    ;

member_initializer_list
    : member_initializer (',' member_initializer)*
    ;

member_initializer
    : initializer_target '=' initializer_value
    ;

initializer_target
    : identifier
    | '[' argument_list ']'
    ;

initializer_value
    : expression
    | object_or_collection_initializer
    ;
```

**More work is needed!**

Nigel-Ecma commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+### §indexable-sequence-general General
+
+An ***indexable sequence*** is an ordered set of zero or more elements having the same type. Any given element can be accessed via an index, and a contiguous subset of elements—referred to as a ***slice***—can be denoted via a range.
```
Ref: #1250 a Proposal to Define Some Container-Related Terms

I think this definition of *indexable sequence* (or *collection*) is not right.

#1250 has a renaming and different definition:

> An ***indexable collection*** is a collection in which any given element can be accessed via an index, and a contiguous subset of elements—referred to as a slice—can be denoted via a range.

#1250 also defines *collection*:

> A ***collection*** is an ordered set of zero or more ***collection element***s that can be used in a polymorphic way. Often, the elements are of the same data type such as `int` or `string`. Sometimes the items derive from a common type; even deriving from the most general type, `object`.

I”m not sure where the “polymorphic” in the *collection* definition comes from, but it does correctly drop the “same type” of *indexable sequence*.

The term “ordered set” allows the reading that it is the *elements* that are ordered (as in [partially] ordered sets and `SortedSet<T>`), which they need not be. It is the indicies that are ordered… usually.

`System.Index` indicies by themselves are actually not ordered and the type does not implement `IComparable<Index>` just `IEquatable<Index>`. They *are* ordered only w.r.t a particular collection, which has some `Length`, this is because indicies can be end-relative.

Maybe:

```suggestion
An ***indexable sequence*** is a set of zero or more elements where any given element can be accessed via an index.

A  ***range*** is a contiguous subset of ordered indices, and the subset of elements accessed via the indices in a range is referred to as a ***slice***.
```

Or something like that…

This is similar to the definition in #1250, but without the underlying definition of collection (for now…).

The definition of range here doesn’t address that a C# `a..b` `Range` does not include `b`, or that a C# `Range` may not be a contiguous set of ordered indices (consider `1..^Int.MaxValue` which is a valid `Range` to construct but useless to use!).

So there is a disparity between the term “range” and C#’s type `Range`– that might be confusing…

jskeet replied:
I'd suggest either "sequence" or "collection" instead of "set" as well. Likewise "subsequence" instead of "subset"?

Nigel-Ecma replied:
@jskeet – Well *subsequence* sounds better than *subcollection*…

Revised suggestion:

```suggestion
An ***indexable sequence*** is a sequence of zero or more elements where any given element can be accessed via an index.

A  ***range*** is a sequence of contiguous ordered indices, and the subsequence of elements accessed via the indices in a range is referred to as a ***slice***.
```

Nigel-Ecma replied:
[Replacement comment, I hit add too quickly]

@jskeet dropping set is a good idea.

But are we taking about *sequences*, which by (English language) definition are ordered, or *collections* which are not?

Do we need to make it clear that any ordering here comes from the index type being ordered, not the elements?

Indices do not need to be ordered *per se*, but a *range* of indices implies ordered indices, and indexable sequence/collection is being defined to be used by ranges.

Which gives us:

```suggestion
An ***indexable sequence*** is a sequence of zero or more elements where any given element can be accessed via an ordered index.

A  ***range*** is a contiguous subset of ordered indices.

A ***slice** is the indexable (sub)sequence of elements selected in index order from an indexable sequence using the indices in a range.
```

Too detailed/convoluted?

jskeet replied:
Question under discussion: does `Dictionary<TKey, TValue>` count as an indexable sequence? (Or rather, do we want it to?)

jskeet replied:
Interesting types:

- `int`
- `System.Index`
- `long` (just for arrays)

We *may* want to limit this to integer-based indexes.

jskeet replied:
How about a `Dictionary<int, string>`?

jskeet replied:
We also had some discussion about what "ordered index" means (especially when `System.Index` isn't inherently ordered).

Nigel-Ecma commented on standard/classes.md:
```diff
@@ -3299,6 +3299,8 @@ In a *ref_property_body* an expression body consisting of `=>` followed by `ref`
 
 When a property declaration includes an `extern` modifier, the property is said to be an ***external property***. Because an external property declaration provides no actual implementation, each of the *accessor_body*s in its *accessor_declarations* shall be a semicolon.
 
+A type is ***countable*** if it has an instance property named `Length` or `Count` with an accessible `get` accessor ([§15.7.3]( classes.md#1573-accessors)) and a return type of `int`.
```
Hmmm…

```
struct Plank
﻿{
   int Length { get; }      // all measurements in mm
   int Width { get; }
   int Thickness { get; }
}
```

A single plank is countable (in the sense intended)?

Maybe:

```suggestion
A collection type is ***countable*** if it has an instance property named `Length` or `Count` with an accessible `get` accessor ([§15.7.3]( classes.md#1573-accessors)) and a return type of `int`.
```

with a x-ref to wherever *collection* gets defined (see #1250).

Note: Qualifying, “collection type”, also avoids integers not be countable – which could otherwise confuse the numerate 😉

Note: This doesn’t address “In case both `Length` and `Count` are present, `Length` will be preferred.” found [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#adding-index-and-range-support-to-existing-library-types).

Nigel-Ecma replied:
While having general terms is good this particular term currently only exists to be referenced in the new clauses “Implicit Index support” and ”Range Index support”. It would be better to move this to down to those two clauses, possibly even removing it as defined term altogether.

jskeet replied:
Let's move this to the general clause of indexable sequences, where we're introducing other terms that are only relevant to that. (This doesn't address the plank example, but at least suggests where countable types are relevant.)

Nigel-Ecma commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+A type having an instance indexer taking a single argument of type `System.Index`, or a first argument of that type followed by optional arguments, may be indexed as described by [§12.8.11.3](expressions.md#128113-indexer-access).
+
+> *Example*: In [§15.9](classes.md#159-indexers), there is an example defining type `BitArray`, which stores bits in an array of `int`. Individual bits are accessed for read/write via an `int` indexer. Adding an `Index` indexer that simply interprets the `Index` argument as an `int`, is simple:
```
Given that the preceding para has, hopefully, just stated the obvious – that adding another indexer is nothing new, I question whether this really needs an example which visually separates the previous para from the next clause that builds upon it.

However I accept this might not be a universal view! 😉

jskeet replied:
If nothing else, we can reduce the clutter using expression bodies:

```csharp
partial class BitArray
{
    public bool this[Index idx]
    {
        get => this[idx.GetOffset(Length)];   // use the [int] indexer
        set => this[idx.GetOffset(Length)] = value;  // use the [int] indexer
    } 
}
```

Nigel-Ecma replied:
@jskeet – that’s an improvement 👍

Nigel-Ecma commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+#### §indexable-sequence-impl-support-for-index Implicit Index support
+
+An implementation shall behave as if it provides a non-virtual instance indexer member with a single parameter of type `System.Index` for any type that meets the following criteria:
+
+- The type is countable ([§15.7.1](classes.md#1571-general)).
+- The type has an accessible instance indexer taking an argument of type `int` as its only argument.
+- The type does not have an accessible instance indexer taking a `System.Index` as its only argument, or as its first argument with the remaining arguments being optional.
```
**Note** See @jnm2 & my previous comments just above – this suggestion is probably going to be wrong but is here just in case it isn’t.

---
Following the suggested take/accept change in the previous clause I suggested:

```suggestion
- The type has an accessible instance indexer accepting an argument of type `int` as its only argument.
- The type does not have an accessible instance indexer accepting a `System.Index` as its only argument.
```

jnm2 replied:
I think by this list, the following should not error, but it does:

```cs
var dict = new X();
dict[^1] = "";

class X : Dictionary<int, string>
{
}
```

Nigel-Ecma commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+The provided instance indexer shall have the same get and set members with matching accessibility as the `int` indexer.
+
+The provided instance indexer shall take the given `System.Index` and use that to call the instance indexer taking an `int`. If both the `Length` and `Count` properties exist and are accessible, `Length` is used.
```
Converting descriptive to specification language:

```suggestion
The provided instance indexer shall take the given `System.Index` and use that to call the instance indexer taking an `int`. If both the `Length` and `Count` properties exist and are accessible, `Length` shall be used.
```

Nigel-Ecma replied:
Further to my last comment, there is no context here as to why either `Length` or `Count` should be used at all – there is no mention of `GetOffset` is there is in expressions.md line 2242. And as commented on the latter, surely we don't need this stuff in two places?

Nigel-Ecma reviewed (changes requested):
I haven’t (yet) considered @BillWagner’s suggestion to rewrite the clauses describing emulating `Index` and `Range` indexing; and I ran out of steam a bit at the end 😉 but I'll post so it’s in before the next meeting.

Quite a few changes, some will be debated or just plain wrong…

jskeet commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+A slice can be obtained using the `..` range operator (§range-operator). A range of the form `s..e` starts at element `s` and ends with the element immediately prior to element `e`.
+
+All single-dimensional and jagged arrays ([§17.1](arrays.md#171-general)) are indexable sequences; multi-dimensional arrays are not! The use of indexes and ranges with arrays is described in [§12.8.11.2](expressions.md#128112-array-access).
```
Should we also say that these arrays are countable as well?

Nigel-Ecma replied:
@jskeet – I’ve now added that to the suggested definition of slice.

jskeet commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+All single-dimensional and jagged arrays ([§17.1](arrays.md#171-general)) are indexable sequences; multi-dimensional arrays are not! The use of indexes and ranges with arrays is described in [§12.8.11.2](expressions.md#128112-array-access).
+
+An object of type `string` is an indexable sequence.
```
And countable? (I won't keep adding this...)

Nigel-Ecma replied:
@jskeet – Once we’ve the definition of indexable narrowed down to what is needed for this feature then countability (if it still exists as a concept) will be implied as using an `Index` relies on it.

jskeet commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+- The type has an accessible instance indexer taking an argument of type `int` as its only argument.
+- The type does not have an accessible instance indexer taking a `System.Index` as its only argument, or as its first argument with the remaining arguments being optional.
+
+The provided instance indexer shall have the same get and set members with matching accessibility as the `int` indexer.
```
I don't really understand what's meant by "the same get and set members" here in the first place... My guess is that this is to do with "if the int-based indexer is readable, then so is the provided Index-based indexer; likewise is the int-based indexer is writable, then so is the provided Index-based indexer" but I could be wrong.

jskeet commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,101 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
 #### 12.8.12.2 Array access
 
-For an array access, the *primary_no_array_creation_expression* of the *element_access* shall be a value of an *array_type*. Furthermore, the *argument_list* of an array access is not allowed to contain named arguments. The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be of type `int`, `uint`, `long`, or `ulong,` or shal ...
+For an array access, the *primary_no_array_creation_expression* of the *element_access* shall be a value of an *array_type*. Furthermore, the *argument_list* of an array access shall not contain named arguments. The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be of type `int`, `uint`, `long`, `ulong`, `System.Index`, or ...
```
It looks like having an expression that's implicitly convertible to `System.Index` or `System.Range` is fine too - but not for multi-dimensional arrays.

Nigel-Ecma replied:
Maybe this is because it’s only doing the conversion for the virtual indexer taking a single argument? See comment a little below – we seem to have competing specs of the behaviour.

Nigel-Ecma replied:
Maybe make this positive by restoring the original para and instead adding a second one; and include @jskeet’s implicitly convertible observation?

```suggestion
For an array access, the *primary_no_array_creation_expression* of the *element_access* shall be a value of an *array_type*. Furthermore, the *argument_list* of an array access shall not contain named arguments. The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be of type `int`, `uint`, `long`, or `ulong`, or shall be implicitly convertible to one or more of these types.

For single-dimension array access the argument expression can also be of type `System.Index` or `System.Range`, or implicitly convertible to one of these.
```

jskeet replied:
Here's an unexpected example:

```csharp
public struct Int32OrRange
{
    public static implicit operator int(Int32OrRange index) => 0;
    public static implicit operator Range(Int32OrRange index) => new Range(0, 0);
}

string[] array = { };
Int32OrRange x = new();
var result = array[x];
```

That compiles, and uses the conversion to `int`. If you remove the conversion, it still compiles, and uses the conversion to `Range`. I'd have expected the first to compile because neither `int` nor `Range` is better than the other.

jskeet commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,101 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
-- The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
-- The value of each expression in the *argument_list* is checked against the actual bounds of each dimension of the array instance referenced by `P`. If one or more values are out of range, a `System.IndexOutOfRangeException` is thrown and no further steps are executed.
-- The location of the array element given by the index expression(s) is computed, and this location becomes the result of the array access.
+- The index expressions of the *argument_list* are evaluated in order, from left to right. Following evaluation of each index expression, for expressions not having type `System.Index` or `System.Range`, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to one of the following types is performed: `int`, `uint`, `long`, `ulong`. The first type in this list for which an impl ...
```
Again, we need to take into account indexer arguments which aren't Index/Range but are convertible to them.

Nigel-Ecma replied:
How does this mesh with the “behave as if” behaviour specified in the new clause *Implicit Index support* (classes.md, line 5594)?

This version at least mentions `GetOffset`, but surely we only need to specify this stuff on one of the places?

Nigel-Ecma replied:
How does this mesh with the “behave as if” behaviour specified in the new clause *Implicit Index support* (classes.md, line 5594)?

This version mentions `GetOffset`, but are we requiring that it be called or that an implementation may “behave as if” it is? Elsewhere in the Standard there are passages where behaviour is defined to be the “equivalent of executing the following code” (or words to that effect), would that pattern be better here?

Nigel-Ecma replied:
I think the valiant attempt to interweave `Index` and `Range` is currently a bit hard to follow, well at least I've had to keep re-untangling it as I check the semantics. I think separating out integral, `Index` and `Range` cases a bit more would help, I might back it back before the meeting to make a suggestion…

jskeet commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,101 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+  - The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
+  - The value of each expression in the *argument_list* is checked against the actual bounds of each dimension of the array instance referenced by `P`. If one or more values are out of range, a `System.IndexOutOfRangeException` is thrown and no further steps are executed.
+  - The location of the array element given by the index expression(s) is computed, and this location becomes the result of the array access.
+- Else, for an index expression having type `System.Range`:
```
Nit: I think we use "Otherwise" typically instead of "Else"

jskeet commented on standard/expressions.md:
````diff
@@ -3557,6 +3645,47 @@ The result of evaluating `~x`, where `X` is an expression of an enumeration ty
+System.Index operator ^(int fromEnd);
+```
+
+For this operator, an object of (the immutable struct) type `System.Index` is returned that denotes element number `fromEnd` from the end of any indexable sequence. `^n` is shorthand for `new System.Index(n, fromEnd: true)`.
````
```suggestion
For this operator, a value of (the immutable struct) type `System.Index` is returned that denotes element number `fromEnd` from the end of any indexable sequence. `^n` is shorthand for `new System.Index(n, fromEnd: true)`.
```

jskeet replied:
(To avoid any suspicion of boxing.)

jskeet commented on standard/expressions.md:
````diff
@@ -3688,6 +3817,55 @@ At run-time, the expression `await t` is evaluated as follows:
+    ;
+```
+
+For an operation of the form `s .. e`, binary operator overload resolution ([§12.4.5](expressions.md#1245-binary-operator-overload-resolution)) is applied to select a specific operator implementation. The operands are converted to the parameter types of the selected operator, and the type of the result is the return type of the operator. *range_expression* shall have type `System.Index` or a type ...
````
```suggestion
For an operation of the form `s .. e`, binary operator overload resolution ([§12.4.5](expressions.md#1245-binary-operator-overload-resolution)) is applied to select a specific operator implementation. The operands are converted to the parameter types of the selected operator, and the type of the result is the return type of the operator. Each of `s` and `e` shall have type `System.Index` or a type that can be converted implicitly to that type. Only one predefined range operator exists:
```

(Or let's clarify in a different way. The fact that `range_expression` uses `range_expression` make this more confusing.)

jskeet commented on standard/expressions.md:
````diff
@@ -3688,6 +3817,55 @@ At run-time, the expression `await t` is evaluated as follows:
+```ANTLR
+range_expression
+    : unary_expression
+    | range_expression? '..' range_expression?
````
Is this definitely correct? That suggests `0 .. 4 .. 5` would be valid, because `0` is a valid range expression, and so is `4 .. 5`.

I think I'm missing something here.

Nigel-Ecma replied:
I haven’t yet taken a look at the grammar, but yes that looks wrong.

jskeet commented on standard/expressions.md:
````diff
@@ -3688,6 +3817,55 @@ At run-time, the expression `await t` is evaluated as follows:
+System.Range operator ..(System.Index start = 0, System.Index end = ^0);
+```
+
+The left and right operands denote, respectively, a start and end Index. For this operator, an object of (the immutable struct) type `System.Range` is returned that contains those Indexes. If the left operand is omitted, an Index of `0` is used. If the right operand is omitted, an Index of `^0` is used. As such,
````
```suggestion
The left and right operands denote, respectively, a start and end Index. For this operator, a value of (the immutable struct) type `System.Range` is returned that contains those Indexes. If the left operand is omitted, an Index of `0` is used. If the right operand is omitted, an Index of `^0` is used. As such,
```

jskeet reviewed (commented):
Stopped reviewing after getting confused at `range_expression`...

Nigel-Ecma commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
 
 This allows the context to keep track of how many `void`-returning async functions are running under it, and to decide how to propagate exceptions coming out of them.
+
+## §indexable-sequence Indexable sequences
```
This clause & its subclauses might be better after §15.9 *Indexers*, or as a subclause of §15.9.

jskeet replied:
Punt on this until the rest is ready.

Nigel-Ecma commented on standard/classes.md:
```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
+
+An index is represented by a read-only variable of the value type `System.Index`. A range is represented by a read-only variable of value type `System.Range`, which contains a start and end index. A slice of an array is represented by a (possibly empty) array. The representation of a slice of a user-defined type is determined by the implementer of that type.
+
+For an indexable sequence of length *N*, elements can be accessed using indexes 0 through *N-1*, which are relative to the start. Elements can also be accessed relative to the end via the `^`  index-from-end operator (§index-from-end-operator). `^0` denotes the (non-existent) element just beyond the end.
```
The definition of *indexable sequence*, both original and revised versions, do not fix the type of the index. Here the type is fixed to being an integer number (it does not state the actual type).

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -348,8 +349,8 @@ In both of the above cases, a cast expression can be used to explicitly convert
 
-- For the unary operators `+`, `++`, `-`, `--`, `!`(logical negation), and `~`, a lifted form of an operator exists if the operand and result types are both non-nullable value types. The lifted form is constructed by adding a single `?` modifier to the operand and result types. The lifted operator produces a `null` value if the operand is `null`. Otherwise, the lifted operator unwraps the operand ...
-- For the binary operators `+`, `-`, `*`, `/`, `%`, `&`, `|`, `^`, `<<`, and `>>`, a lifted form of an operator exists if the operand and result types are all non-nullable value types. The lifted form is constructed by adding a single `?` modifier to each operand and result type. The lifted operator produces a `null` value if one or both operands are `null` (an exception being the `&` and `|` ope ...
+- For the unary operators `+`, `++`, `-`, `--`, `!`, `~`, and `^`, a lifted form of an operator exists if the operand and result types are both non-nullable value types. The lifted form is constructed by adding a single `?` modifier to the operand and result types. The lifted operator produces a `null` value if the operand is `null`. Otherwise, the lifted operator unwraps the operand, applies the ...
```
The change removed the unconnected “(logical negation)” qualification on `!` which exists to distinguish it from the postfix null-forgiving operator and should be restored.

```suggestion
- For the unary operators `+`, `++`, `-`, `--`, `!`(logical negation), `~`, and `^`, a lifted form of an operator exists if the operand and result types are both non-nullable value types. The lifted form is constructed by adding a single `?` modifier to the operand and result types. The lifted operator produces a `null` value if the operand is `null`. Otherwise, the lifted operator unwraps the operand, applies the underlying operator, and wraps the result.
```

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,101 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+
+- For an index expression not having type `System.Range`:
+  - If evaluation of an index expression, the subsequent implicit conversion, or Index transformation causes an exception, then no further index expressions are evaluated, and no further steps are executed.
+  - The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
```
This check on `P` being valid is occurring *after* the conversion of `Index` to `int` which requires `P`…

jnm2 commented on standard/standard-library.md:
```diff
@@ -379,18 +379,44 @@ The following types, including the members listed, shall be defined in a conform
+        public override bool Equals(object? value);
+        public bool Equal (Range other);
+        public override int GetHashCode();
+        public (int,int) GetOffsetAndLength(int length);
```
Should we mention the tuple element names that the BCL shipped?
```suggestion
        public (int Offset, int Length) GetOffsetAndLength(int length);
```
(<https://learn.microsoft.com/en-us/dotnet/api/system.range.getoffsetandlength?view=netcore-3.0>)

jskeet committed:
Update standard/expressions.md

Co-authored-by: Nigel-Ecma <6654683+Nigel-Ecma@users.noreply.github.com>

jskeet committed:
Update standard/expressions.md

Co-authored-by: Nigel-Ecma <6654683+Nigel-Ecma@users.noreply.github.com>

jskeet commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,103 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+For an array access, the *primary_no_array_creation_expression* of the *element_access* shall be a value of an *array_type*. Furthermore, the *argument_list* of an array access shall not contain named arguments. The number of expressions in the *argument_list* shall be the same as the rank of the *array_type*, and each expression shall be of type `int`, `uint`, `long`, or `ulong`, or shall be imp ...
 
-The result of evaluating an array access is a variable of the element type of the array, namely the array element selected by the value(s) of the expression(s) in the *argument_list*.
+For single-dimension array access the argument expression can also be of type `System.Index` or `System.Range`, or implicitly convertible to one of these.
```
Here's an unexpected example:

```csharp
public struct Int32OrRange
{
    public static implicit operator int(Int32OrRange index) => 0;
    public static implicit operator Range(Int32OrRange index) => new Range(0, 0);
}

string[] array = { };
Int32OrRange x = new();
var result = array[x];
```

That compiles, and uses the conversion to int. If you remove the conversion, it still compiles, and uses the conversion to Range. I'd have expected the first to compile because neither `int` nor `Range` is better than the other.

BillWagner replied:
I checked with the compiler team.  The roslyn implementation ranks single argument indexers in the following order:

- int
- uint
- Int64
- UInt64
- Index
- Range

I wouldn't expect that to change, so we'll have to come up with normative rules that match that behavior.

jnm2 commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,103 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+
+The result of evaluating an array access that does not involve `System.Range` is a variable of the element type of the array, namely the array element selected by the value(s) of the expression(s) in the *argument_list*.
+
+The result of evaluating an array access that involves `System.Range` is a slice (§indexable-sequence-general) of the array being accessed, as selected by the value(s) of the expression(s) in the *argument_list*. The resulting slice’s element type is the same as the element type of the array being accessed.
```
Can we specify that the returned slice must be an array type?
```suggestion
The result of evaluating an array access that involves `System.Range` is a slice (§indexable-sequence-general) of the array being accessed, as selected by the value(s) of the expression(s) in the *argument_list*. The resulting slice has the same type as the array being accessed, both at compile time and runtime.
```

jskeet commented on standard/expressions.md:
```diff
@@ -2227,17 +2228,103 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+
+The result of evaluating an array access that does not involve `System.Range` is a variable of the element type of the array, namely the array element selected by the value(s) of the expression(s) in the *argument_list*.
+
+The result of evaluating an array access that involves `System.Range` is a slice (§indexable-sequence-general) of the array being accessed, as selected by the value(s) of the expression(s) in the *argument_list*. The resulting slice’s element type is the same as the element type of the array being accessed.
```
We may need to differentiate between compile-time and runtime types here. Does the code below throw an exception?

```csharp
object[] obj = new string[5];
object[] slice = obj[0..5];
slice[0] = new object();
```

jskeet replied:
(Spoiler alert: it throws an exception.)

jskeet replied:
This concurs with Joseph's suggestion, for the removal of any doubt.

BillWagner replied:
And this example prints "System.String[]":

```
string[] source = ["one", "two", "three", "four"] ;

object[] basic = source;

var slice = basic[0..2];
Type t = slice.GetType();
Console.WriteLine(t);
```

jskeet committed:
Update standard/expressions.md

Co-authored-by: Joseph Musser <me@jnm2.com>

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -3688,6 +3819,55 @@ At run-time, the expression `await t` is evaluated as follows:
+
+The left and right operands denote, respectively, a start and end Index. For this operator, an object of (the immutable struct) type `System.Range` is returned that contains those Indexes. If the left operand is omitted, an Index of `0` is used. If the right operand is omitted, an Index of `^0` is used. As such,
+
+- `s .. e` is transformed by the implementation to `new System.Range(s, e)`.
+- `s ..` is transformed by the implementation to `new System.Range(s, ^0)`.
+- `.. e` is transformed by the implementation to `new System.Range(0, e)`.
+- `..` is transformed by the implementation to `new System.Range(0, ^0)`.
```
“transformed by the implementation” is wrong here as:
- there is no requirement that the implementation does this transformation (and Roslyn does not always do so); and
- the Standard uses phrasing like “has the same meaning as” or “is equivalent to” (e.g. §12.8.8, §13.9.5)

jskeet commented:
@Nigel-Ecma IIRC you were going to propose a new structure for this (in a new PR, I believe?) - do you think you'll have anything ready in time for us to review by the next meeting? Do let me know if I'm completely wrong on this.

Nigel-Ecma commented:
> @Nigel-Ecma IIRC you were going to propose a new structure for this (in a new PR, I believe?) - do you think you'll have anything ready in time for us to review by the next meeting? Do let me know if I'm completely wrong on this.

@jskeet – You are correct, but I doubt I’ll have anything ready for the next meeting on this one.

RexJaeschke commented:
For the next TG2 call agenda, I propose the following two 15-minute sessions, in order, for this PR, even if Nigel won't have a formal presentation ready:

1. We identify Nigel's main concerns, so we all understand them and can work on addressing them in the weeks following.
2. Separately, we walk through the "File changes" and deal with as many non-controversial corrections and suggestions (lexical, maybe library, and some of the expressions chapter). Once that is done, perhaps we can move the remaining open commits to a new PR, so going forward we don't have to wade through all the comments we just closed/adopted.

jskeet commented:
Removing the "meeting: priority" label as #1369 is now the proposal.

BillWagner commented:
Closed in favor of #1369
